SCSH_3C_iOS/Persistence/CoreData/Board/BoardCoreDataService.swift

```diff
@@ -0,0 +1,152 @@
+            guard let self = self else { return }
+            
+            promise(self.save(board: board))
+        }
+        .eraseToAnyPublisher()
+    }
+
+    public func saveAll(boards: [Board]) -> AnyPublisher<Void, Error> {
+        Future<Void, Error> { [weak self] promise in
+            guard let self = self else { return }
+            
+            promise(self.saveAll(boards: boards))
+        }
+        .eraseToAnyPublisher()
+    }
+    
+    private func get(code: String?, context: NSManagedObjectContext = CoreDataController.shared.container.newBackgroundContext()) -> Result<[Board], Error> {
+        let fetchRequest = Board.fetchRequest()
+        fetchRequest.sortDescriptors = [NSSortDescriptor(key: "id", ascending: true)]
+        
+        if let code = code {
+            fetchRequest.predicate = NSPredicate(format: "code == %@", code)
+        }
+        
+        do {
+            let results = try context.fetch(fetchRequest)
+            try context.save()
+            
+            return .success(results)
+            
+        } catch {
+            return .failure(error)
+        }
+    }
+    
+    private func save(board: Board) -> Result<Void, Error> {
+        guard let code = board.code else {
+            // TODO: Error handling
+            return .failure(NSError(domain: "", code: 0))
+        }
+        let context = board.managedObjectContext ?? CoreDataController.shared.container.newBackgroundContext()
+        let result: Result<[Board], Error> = get(code: code)
+
+        switch result {
+        case .success(let existingBoards):
+            
+            if let existingBoard = existingBoards.first {
+                update(existingBoard: existingBoard, newBoard: board)
+                context.delete(board)
+            } else {
+                context.insert(board)
+            }
+
+            do {
+                try context.save()
+
+                return .success(())
+            } catch {
+                print("❌❌❌ BoardCoreDataService saveAll save fail !")
+                print(error.localizedDescription)
+
+                return .failure(error)
+            }
+            
+        case .failure(let error):
+            print("❌❌❌ BoardCoreDataService saveAll get fail !")
+            print(error.localizedDescription)
+            
+            return .failure(error)
+        }
+    }
+    
+    private func update(existingBoard: Board, newBoard: Board) {
+        existingBoard.id = newBoard.id
+        existingBoard.code = newBoard.code
+        existingBoard.action_type = newBoard.action_type
+        existingBoard.action = newBoard.action
+        existingBoard.image_url = newBoard.image_url
+    }
+    
+    private func saveAll(boards: [Board]) -> Result<Void, Error> {
+        
+        let context = boards.first?.managedObjectContext ?? CoreDataController.shared.container.newBackgroundContext()
+        
+        truncate(context: context)
```

Comment:
> I'm curious why errors aren't handled here then returning an failure early .

I can't find where the code corresponds to the situation you're referring to, Do you mean func `truncate` should be inside the scope of `do catch` in the func `saveAll` ?

Reply:
OK, maybe I misunderstood the try catch pattern in Swift.

So, you just want to omit the error from `truncate` there? or we can say the error of `truncate` isn't important here.

Reply:
> So, you just want to omit the error from truncate there? or we can say the error of truncate isn't important here.

No, if the `truncate` operation fails, it could result in incorrect data in the persistence layer (such as duplicated data).

You're right, I see what you mean, The `truncate` should indeed be inside the `do catch` block in the `saveAll` function.

It's a BUG, thanks for checking for me !!, really appreciate it, I'll fix it later and check other places.

Reply:
https://github.com/No-57/SCSH_3C_iOS/pull/45/commits/01c49b61fd3aec86bf0212fbb8aa2f4df7f3c718

it's a shame that, based on the current design, I'm unable to mock a 'truncate fails' scenario to demonstrate this in the test case, I might consider refactoring this object once I have time.

Reply:
That's fine, take your time.